[CI] TransformIT testStopWaitForCheckpoint failing

**Build scan:**
https://gradle-enterprise.elastic.co/s/q6kyadnnxpljg/tests/:x-pack:plugin:transform:qa:multi-node-tests:javaRestTest/org.elasticsearch.xpack.transform.integration.TransformIT/testStopWaitForCheckpoint

**Reproduction line:**
`./gradlew ':x-pack:plugin:transform:qa:multi-node-tests:javaRestTest' --tests "org.elasticsearch.xpack.transform.integration.TransformIT.testStopWaitForCheckpoint" -Dtests.seed=545D076A9340F142 -Dtests.locale=id -Dtests.timezone=Australia/Victoria -Druntime.java=8`

**Applicable branches:**
7.13, 7.x

**Reproduces locally?:**
Didn't try

**Failure history:**
https://gradle-enterprise.elastic.co/scans/tests?tests.container=org.elasticsearch.xpack.transform.integration.TransformIT&tests.test=testStopWaitForCheckpoint

**Failure excerpt:**
```
org.elasticsearch.ElasticsearchStatusException: Elasticsearch exception [type=status_exception, reason=Failed to update transform task [transform-wait-for-checkpoint] state value should_stop_at_checkpoint from [true] to [true]]

  at __randomizedtesting.SeedInfo.seed([545D076A9340F142:29882B015FD41B74]:0)
  at org.elasticsearch.rest.BytesRestResponse.errorFromXContent(BytesRestResponse.java:176)
  at org.elasticsearch.client.RestHighLevelClient.parseEntity(RestHighLevelClient.java:1933)
  at org.elasticsearch.client.RestHighLevelClient.parseResponseException(RestHighLevelClient.java:1910)
  at org.elasticsearch.client.RestHighLevelClient.internalPerformRequest(RestHighLevelClient.java:1667)
  at org.elasticsearch.client.RestHighLevelClient.performRequest(RestHighLevelClient.java:1639)
  at org.elasticsearch.client.RestHighLevelClient.performRequestAndParseEntity(RestHighLevelClient.java:1606)
  at org.elasticsearch.client.TransformClient.stopTransform(TransformClient.java:308)
  at org.elasticsearch.xpack.transform.integration.TransformIntegTestCase.stopTransform(TransformIntegTestCase.java:145)
  at org.elasticsearch.xpack.transform.integration.TransformIT.testStopWaitForCheckpoint(TransformIT.java:302)
  at sun.reflect.NativeMethodAccessorImpl.invoke0(NativeMethodAccessorImpl.java:-2)
  at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
  at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
  at java.lang.reflect.Method.invoke(Method.java:498)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1750)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:938)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:974)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:988)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:49)
  at org.apache.lucene.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:45)
  at org.apache.lucene.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:48)
  at org.apache.lucene.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:64)
  at org.apache.lucene.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:47)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:368)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:817)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:468)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:947)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:832)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:883)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:894)
  at org.apache.lucene.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:45)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:41)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
  at org.apache.lucene.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:47)
  at org.apache.lucene.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:64)
  at org.apache.lucene.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:54)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:368)
  at java.lang.Thread.run(Thread.java:748)

```

After analyzing the logs, it seems this is caused by quickly starting the transform after stopping it.

The transform is stopped, but the state hasn't been persisted yet:
```
[2021-08-02T05:07:00,272][DEBUG][o.e.x.c.i.AsyncTwoPhaseIndexer] [javaRestTest-2] Finished indexing for job [transform-wait-for-checkpoint], saving state and shutting down.
[2021-08-02T05:07:00,305][DEBUG][o.e.x.t.t.TransformIndexer] [javaRestTest-2] [transform-wait-for-checkpoint] finished indexing for transform checkpoint [4].
[2021-08-02T05:07:00,307][INFO ][o.e.x.t.t.TransformIndexer] [javaRestTest-2] [transform-wait-for-checkpoint] transform has stopped.
[2021-08-02T05:07:00,308][DEBUG][o.e.x.t.t.ClientTransformIndexer] [javaRestTest-2] [transform-wait-for-checkpoint] updating persistent state of transform to [{"task_state":"stopped","indexer_state":"stopped","checkpoint":4,"progress":{"docs_indexed":25,"docs_processed":25},"should_stop_at_checkpoint":false}].
```

Now the transform is started again, and it loads an old state document:

```
[2021-08-02T05:07:00,319][DEBUG][o.e.x.t.t.TransformPersistentTasksExecutor] [javaRestTest-2] [transform-wait-for-checkpoint] Loading existing state: [{"task_state":"stopped","indexer_state":"stopped","checkpoint":3,"should_stop_at_checkpoint":false}], position [null]
```

It loaded checkpoint 3, but it should be 4.

The indexer got initialized with _wrong_ state:
```
[2021-08-02T05:07:00,327][DEBUG][o.e.x.t.t.TransformTask  ] [javaRestTest-2] [transform-wait-for-checkpoint] start called with state [{"task_state":"stopped","indexer_state":"stopped","checkpoint":0,"should_stop_at_checkpoint":false}].
[2021-08-02T05:07:00,328][INFO ][o.e.x.t.t.TransformTask  ] [javaRestTest-2] [transform-wait-for-checkpoint] updating state for transform to [{"task_state":"started","indexer_state":"stopped","checkpoint":3,"should_stop_at_checkpoint":false}].
```

Now, finally the state with checkpoint 4 returned:

```
[2021-08-02T05:07:00,336][DEBUG][o.e.x.t.t.ClientTransformIndexer] [javaRestTest-2] [transform-wait-for-checkpoint] Updated state document from [{seqNo=10,primaryTerm=1,index=.transform-internal-007}] to [{seqNo=12,primaryTerm=1,index=.transform-internal-007}]
```

however, that's the old indexer, the new one has the old state and therefore stumbles upon saving:

```
[2021-08-02T05:07:00,443][ERROR][o.e.x.t.t.ClientTransformIndexer] [javaRestTest-2] [transform-wait-for-checkpoint] updating stats of transform failed.
java.lang.RuntimeException: Failed to persist transform statistics for transform [transform-wait-for-checkpoint]
	at org.elasticsearch.xpack.transform.persistence.IndexBasedTransformConfigManager.lambda$putOrUpdateTransformStoredDoc$14(IndexBasedTransformConfigManager.java:530) ~[?:?]
	at org.elasticsearch.action.ActionListener$1.onFailure(ActionListener.java:142) ~[elasticsearch-7.13.5-SNAPSHOT.jar:7.13.5-SNAPSHOT]
	at org.elasticsearch.action.support.ContextPreservingActionListener.onFailure(ContextPreservingActionListener.java:38) ~[elasticsearch-7.13.5-SNAPSHOT.jar:7.13.5-SNAPSHOT]
	at org.elasticsearch.action.support.TransportAction$1.onFailure(TransportAction.java:92) ~[elasticsearch-7.13.5-SNAPSHOT.jar:7.13.5-SNAPSHOT]
	at org.elasticsearch.action.support.ContextPreservingActionListener.onFailure(ContextPreservingActionListener.java:38) ~[elasticsearch-7.13.5-SNAPSHOT.jar:7.13.5-SNAPSHOT]
	at org.elasticsearch.action.ActionListener$Delegating.onFailure(ActionListener.java:66) ~[elasticsearch-7.13.5-SNAPSHOT.jar:7.13.5-SNAPSHOT]
	at org.elasticsearch.action.bulk.TransportSingleItemBulkWriteAction.lambda$wrapBulkResponse$0(TransportSingleItemBulkWriteAction.java:53) ~[elasticsearch-7.13.5-SNAPSHOT.jar:7.13.5-SNAPSHOT]
	at org.elasticsearch.action.ActionListener$1.onResponse(ActionListener.java:134) ~[elasticsearch-7.13.5-SNAPSHOT.jar:7.13.5-SNAPSHOT]
	at org.elasticsearch.action.support.ContextPreservingActionListener.onResponse(ContextPreservingActionListener.java:31) ~[elasticsearch-7.13.5-SNAPSHOT.jar:7.13.5-SNAPSHOT]
	at org.elasticsearch.xpack.security.action.filter.SecurityActionFilter.lambda$applyInternal$2(SecurityActionFilter.java:163) ~[?:?]
	at org.elasticsearch.action.ActionListener$DelegatingFailureActionListener.onResponse(ActionListener.java:217) ~[elasticsearch-7.13.5-SNAPSHOT.jar:7.13.5-SNAPSHOT]
	at org.elasticsearch.action.ActionListener$RunBeforeActionListener.onResponse(ActionListener.java:387) ~[elasticsearch-7.13.5-SNAPSHOT.jar:7.13.5-SNAPSHOT]
	at org.elasticsearch.action.bulk.TransportBulkAction$BulkOperation$1.finishHim(TransportBulkAction.java:538) ~[elasticsearch-7.13.5-SNAPSHOT.jar:7.13.5-SNAPSHOT]
	at org.elasticsearch.action.bulk.TransportBulkAction$BulkOperation$1.onResponse(TransportBulkAction.java:519) ~[elasticsearch-7.13.5-SNAPSHOT.jar:7.13.5-SNAPSHOT]
	at org.elasticsearch.action.bulk.TransportBulkAction$BulkOperation$1.onResponse(TransportBulkAction.java:508) ~[elasticsearch-7.13.5-SNAPSHOT.jar:7.13.5-SNAPSHOT]
	at org.elasticsearch.action.support.TransportAction$1.onResponse(TransportAction.java:83) ~[elasticsearch-7.13.5-SNAPSHOT.jar:7.13.5-SNAPSHOT]
	at org.elasticsearch.action.support.TransportAction$1.onResponse(TransportAction.java:77) ~[elasticsearch-7.13.5-SNAPSHOT.jar:7.13.5-SNAPSHOT]
	at org.elasticsearch.action.support.ContextPreservingActionListener.onResponse(ContextPreservingActionListener.java:31) ~[elasticsearch-7.13.5-SNAPSHOT.jar:7.13.5-SNAPSHOT]
	at org.elasticsearch.xpack.security.action.filter.SecurityActionFilter.lambda$applyInternal$2(SecurityActionFilter.java:163) ~[?:?]
	at org.elasticsearch.action.ActionListener$DelegatingFailureActionListener.onResponse(ActionListener.java:217) ~[elasticsearch-7.13.5-SNAPSHOT.jar:7.13.5-SNAPSHOT]
	at org.elasticsearch.action.support.replication.TransportReplicationAction$ReroutePhase.finishOnSuccess(TransportReplicationAction.java:877) ~[elasticsearch-7.13.5-SNAPSHOT.jar:7.13.5-SNAPSHOT]
	at org.elasticsearch.action.support.replication.TransportReplicationAction$ReroutePhase$1.handleResponse(TransportReplicationAction.java:796) ~[elasticsearch-7.13.5-SNAPSHOT.jar:7.13.5-SNAPSHOT]
	at org.elasticsearch.action.support.replication.TransportReplicationAction$ReroutePhase$1.handleResponse(TransportReplicationAction.java:787) ~[elasticsearch-7.13.5-SNAPSHOT.jar:7.13.5-SNAPSHOT]
	at org.elasticsearch.transport.TransportService$5.handleResponse(TransportService.java:732) ~[elasticsearch-7.13.5-SNAPSHOT.jar:7.13.5-SNAPSHOT]
	at org.elasticsearch.transport.TransportService$ContextRestoreResponseHandler.handleResponse(TransportService.java:1273) ~[elasticsearch-7.13.5-SNAPSHOT.jar:7.13.5-SNAPSHOT]
	at org.elasticsearch.transport.InboundHandler.doHandleResponse(InboundHandler.java:291) ~[elasticsearch-7.13.5-SNAPSHOT.jar:7.13.5-SNAPSHOT]
	at org.elasticsearch.transport.InboundHandler.handleResponse(InboundHandler.java:275) ~[elasticsearch-7.13.5-SNAPSHOT.jar:7.13.5-SNAPSHOT]
	at org.elasticsearch.transport.InboundHandler.messageReceived(InboundHandler.java:128) ~[elasticsearch-7.13.5-SNAPSHOT.jar:7.13.5-SNAPSHOT]
	at org.elasticsearch.transport.InboundHandler.inboundMessage(InboundHandler.java:84) ~[elasticsearch-7.13.5-SNAPSHOT.jar:7.13.5-SNAPSHOT]
	at org.elasticsearch.transport.TcpTransport.inboundMessage(TcpTransport.java:693) ~[elasticsearch-7.13.5-SNAPSHOT.jar:7.13.5-SNAPSHOT]
	at org.elasticsearch.transport.InboundPipeline.forwardFragments(InboundPipeline.java:129) [elasticsearch-7.13.5-SNAPSHOT.jar:7.13.5-SNAPSHOT]
	at org.elasticsearch.transport.InboundPipeline.doHandleBytes(InboundPipeline.java:104) [elasticsearch-7.13.5-SNAPSHOT.jar:7.13.5-SNAPSHOT]
	at org.elasticsearch.transport.InboundPipeline.handleBytes(InboundPipeline.java:69) [elasticsearch-7.13.5-SNAPSHOT.jar:7.13.5-SNAPSHOT]
	at org.elasticsearch.transport.netty4.Netty4MessageChannelHandler.channelRead(Netty4MessageChannelHandler.java:63) [transport-netty4-client-7.13.5-SNAPSHOT.jar:7.13.5-SNAPSHOT]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) [netty-transport-4.1.49.Final.jar:4.1.49.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) [netty-transport-4.1.49.Final.jar:4.1.49.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) [netty-transport-4.1.49.Final.jar:4.1.49.Final]
	at io.netty.handler.logging.LoggingHandler.channelRead(LoggingHandler.java:271) [netty-handler-4.1.49.Final.jar:4.1.49.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) [netty-transport-4.1.49.Final.jar:4.1.49.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) [netty-transport-4.1.49.Final.jar:4.1.49.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) [netty-transport-4.1.49.Final.jar:4.1.49.Final]
	at io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1518) [netty-handler-4.1.49.Final.jar:4.1.49.Final]
	at io.netty.handler.ssl.SslHandler.decodeJdkCompatible(SslHandler.java:1267) [netty-handler-4.1.49.Final.jar:4.1.49.Final]
	at io.netty.handler.ssl.SslHandler.decode(SslHandler.java:1314) [netty-handler-4.1.49.Final.jar:4.1.49.Final]
	at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:501) [netty-codec-4.1.49.Final.jar:4.1.49.Final]
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:440) [netty-codec-4.1.49.Final.jar:4.1.49.Final]
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:276) [netty-codec-4.1.49.Final.jar:4.1.49.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) [netty-transport-4.1.49.Final.jar:4.1.49.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) [netty-transport-4.1.49.Final.jar:4.1.49.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) [netty-transport-4.1.49.Final.jar:4.1.49.Final]
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410) [netty-transport-4.1.49.Final.jar:4.1.49.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) [netty-transport-4.1.49.Final.jar:4.1.49.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) [netty-transport-4.1.49.Final.jar:4.1.49.Final]
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919) [netty-transport-4.1.49.Final.jar:4.1.49.Final]
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:163) [netty-transport-4.1.49.Final.jar:4.1.49.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:714) [netty-transport-4.1.49.Final.jar:4.1.49.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysPlain(NioEventLoop.java:615) [netty-transport-4.1.49.Final.jar:4.1.49.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:578) [netty-transport-4.1.49.Final.jar:4.1.49.Final]
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493) [netty-transport-4.1.49.Final.jar:4.1.49.Final]
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989) [netty-common-4.1.49.Final.jar:4.1.49.Final]
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) [netty-common-4.1.49.Final.jar:4.1.49.Final]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_301]
Caused by: org.elasticsearch.index.engine.VersionConflictEngineException: [data_frame_transform_state_and_stats-transform-wait-for-checkpoint]: version conflict, required seqNo [10], primary term [1]. current document has seqNo [12] and primary term [1]
```

Summary: It is possible to start a transform, although the state hasn't been saved yet / the former `stop` hasn't been finished.

--- a/x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/action/TransportStopTransformAction.java
+++ b/x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/action/TransportStopTransformAction.java
@@ -256,6 +256,7 @@ public class TransportStopTransformAction extends TransportTasksAction<Transform
                     new ElasticsearchStatusException(
                         "Failed to update transform task [{}] state value should_stop_at_checkpoint from [{}] to [{}]",
                         RestStatus.CONFLICT,
+                        e,
                         transformTask.getTransformId(),
                         transformTask.getState().shouldStopAtNextCheckpoint(),
                         request.isWaitForCheckpoint()

--- a/x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/action/TransportStopTransformAction.java
+++ b/x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/action/TransportStopTransformAction.java
@@ -244,25 +244,28 @@ public class TransportStopTransformAction extends TransportTasksAction<Transform
         }
 
         if (ids.contains(transformTask.getTransformId())) {
-            transformTask.setShouldStopAtCheckpoint(request.isWaitForCheckpoint(), ActionListener.wrap(r -> {
-                try {
-                    transformTask.stop(request.isForce(), request.isWaitForCheckpoint());
-                    listener.onResponse(new Response(true));
-                } catch (ElasticsearchException ex) {
-                    listener.onFailure(ex);
-                }
-            },
-                e -> listener.onFailure(
-                    new ElasticsearchStatusException(
-                        "Failed to update transform task [{}] state value should_stop_at_checkpoint from [{}] to [{}]",
-                        RestStatus.CONFLICT,
-                        e,
-                        transformTask.getTransformId(),
-                        transformTask.getState().shouldStopAtNextCheckpoint(),
-                        request.isWaitForCheckpoint()
+            // move the call to the generic thread pool, so we do not block the network thread
+            threadPool.executor(ThreadPool.Names.GENERIC).execute(() -> {
+                transformTask.setShouldStopAtCheckpoint(request.isWaitForCheckpoint(), ActionListener.wrap(r -> {
+                    try {
+                        transformTask.stop(request.isForce(), request.isWaitForCheckpoint());
+                        listener.onResponse(new Response(true));
+                    } catch (ElasticsearchException ex) {
+                        listener.onFailure(ex);
+                    }
+                },
+                    e -> listener.onFailure(
+                        new ElasticsearchStatusException(
+                            "Failed to update transform task [{}] state value should_stop_at_checkpoint from [{}] to [{}]",
+                            RestStatus.CONFLICT,
+                            e,
+                            transformTask.getTransformId(),
+                            transformTask.getState().shouldStopAtNextCheckpoint(),
+                            request.isWaitForCheckpoint()
+                        )
                     )
-                )
-            ));
+                ));
+            });
         } else {
             listener.onFailure(
                 new RuntimeException(

--- a/x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/ClientTransformIndexer.java
+++ b/x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/ClientTransformIndexer.java
@@ -362,6 +362,12 @@ class ClientTransformIndexer extends TransformIndexer {
         }));
     }
 
+    /**
+     * Runs the persistent part of state storage
+     *
+     * Important: Might call into TransformTask, this should _only_ be called with an acquired indexer lock if and only if
+     * the lock for TransformTask has been acquired, too. See gh#75846
+     */
     private void doSaveState(TransformState state, ActionListener<Void> listener) {
         // This could be `null` but the putOrUpdateTransformStoredDoc handles that case just fine
         SeqNoPrimaryTermAndIndex seqNoPrimaryTermAndIndex = getSeqNoPrimaryTermAndIndex();

--- a/x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/ClientTransformIndexer.java
+++ b/x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/ClientTransformIndexer.java
@@ -56,7 +56,6 @@ import org.elasticsearch.xpack.transform.persistence.SeqNoPrimaryTermAndIndex;
 import org.elasticsearch.xpack.transform.transforms.pivot.SchemaUtil;
 import org.elasticsearch.xpack.transform.utils.ExceptionRootCauseFinder;
 
-import java.util.Collection;
 import java.util.LinkedHashMap;
 import java.util.Map;
 import java.util.Map.Entry;
@@ -255,119 +254,6 @@ class ClientTransformIndexer extends TransformIndexer {
             getConfig().getDestination().getIndex(),
             fieldMappingsListener
         );
-    }
-
-    @Override
-    protected void doSaveState(IndexerState indexerState, TransformIndexerPosition position, Runnable next) {
-        if (context.getTaskState() == TransformTaskState.FAILED) {
-            logger.debug("[{}] attempted to save state and stats while failed.", getJobId());
-            // If we are failed, we should call next to allow failure handling to occur if necessary.
-            next.run();
-            return;
-        }
-        if (indexerState.equals(IndexerState.ABORTING)) {
-            // If we're aborting, just invoke `next` (which is likely an onFailure handler)
-            next.run();
-            return;
-        }
-
-        // getting the listeners that registered till now, in theory a new listener could sneak in between this line
-        // and the next, however this is benign: we would store `shouldStopAtCheckpoint = true` twice which is ok
-        Collection<ActionListener<Void>> saveStateListenersAtTheMomentOfCalling = saveStateListeners.getAndSet(null);
-        boolean shouldStopAtCheckpoint = context.shouldStopAtCheckpoint();
-
-        // If we should stop at the next checkpoint, are STARTED, and with `initialRun()` we are in one of two states
-        // 1. We have just called `onFinish` completing our request, but `shouldStopAtCheckpoint` was set to `true` before our check
-        // there and now
-        // 2. We are on the very first run of a NEW checkpoint and got here either through a failure, or the very first save state call.
-        //
-        // In either case, we should stop so that we guarantee a consistent state and that there are no partially completed checkpoints
-        if (shouldStopAtCheckpoint && initialRun() && indexerState.equals(IndexerState.STARTED)) {
-            indexerState = IndexerState.STOPPED;
-            auditor.info(transformConfig.getId(), "Transform is no longer in the middle of a checkpoint, initiating stop.");
-            logger.info("[{}] transform is no longer in the middle of a checkpoint, initiating stop.", transformConfig.getId());
-        }
-
-        // This means that the indexer was triggered to discover changes, found none, and exited early.
-        // If the state is `STOPPED` this means that TransformTask#stop was called while we were checking for changes.
-        // Allow the stop call path to continue
-        if (hasSourceChanged == false && indexerState.equals(IndexerState.STOPPED) == false) {
-            if (saveStateListenersAtTheMomentOfCalling != null) {
-                ActionListener.onResponse(saveStateListenersAtTheMomentOfCalling, null);
-            }
-            next.run();
-            return;
-        }
-
-        TransformTaskState taskState = context.getTaskState();
-
-        if (indexerState.equals(IndexerState.STARTED) && context.getCheckpoint() == 1 && this.isContinuous() == false) {
-            // set both to stopped so they are persisted as such
-            indexerState = IndexerState.STOPPED;
-
-            auditor.info(transformConfig.getId(), "Transform finished indexing all data, initiating stop");
-            logger.info("[{}] transform finished indexing all data, initiating stop.", transformConfig.getId());
-        }
-
-        // If we are `STOPPED` on a `doSaveState` call, that indicates we transitioned to `STOPPED` from `STOPPING`
-        // OR we called `doSaveState` manually as the indexer was not actively running.
-        // Since we save the state to an index, we should make sure that our task state is in parity with the indexer state
-        if (indexerState.equals(IndexerState.STOPPED)) {
-            // If we are going to stop after the state is saved, we should NOT persist `shouldStopAtCheckpoint: true` as this may
-            // cause problems if the task starts up again.
-            // Additionally, we don't have to worry about inconsistency with the ClusterState (if it is persisted there) as the
-            // when we stop, we mark the task as complete and that state goes away.
-            shouldStopAtCheckpoint = false;
-
-            // We don't want adjust the stored taskState because as soon as it is `STOPPED` a user could call
-            // .start again.
-            taskState = TransformTaskState.STOPPED;
-        }
-
-        final TransformState state = new TransformState(
-            taskState,
-            indexerState,
-            position,
-            context.getCheckpoint(),
-            context.getStateReason(),
-            getProgress(),
-            null,
-            shouldStopAtCheckpoint
-        );
-        logger.debug("[{}] updating persistent state of transform to [{}].", transformConfig.getId(), state.toString());
-
-        // we might need to call the save state listeners, but do not want to stop rolling
-        persistStateWithAutoStop(state, ActionListener.wrap(r -> {
-            try {
-                if (saveStateListenersAtTheMomentOfCalling != null) {
-                    ActionListener.onResponse(saveStateListenersAtTheMomentOfCalling, r);
-                }
-            } catch (Exception onResponseException) {
-                String msg = LoggerMessageFormat.format("[{}] failed notifying saveState listeners, ignoring.", getJobId());
-                logger.warn(msg, onResponseException);
-            } finally {
-                next.run();
-            }
-        }, e -> {
-            try {
-                if (saveStateListenersAtTheMomentOfCalling != null) {
-                    ActionListener.onFailure(saveStateListenersAtTheMomentOfCalling, e);
-                }
-            } catch (Exception onFailureException) {
-                String msg = LoggerMessageFormat.format("[{}] failed notifying saveState listeners, ignoring.", getJobId());
-                logger.warn(msg, onFailureException);
-            } finally {
-                next.run();
-            }
-        }));
-    }
-
-    private void persistStateWithAutoStop(TransformState state, ActionListener<Void> listener) {
-        persistState(state, ActionListener.runBefore(listener, () -> {
-            if (state.getTaskState().equals(TransformTaskState.STOPPED)) {
-                context.shutdown();
-            }
-        }));
     }
 
     /**

--- a/x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/ClientTransformIndexer.java
+++ b/x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/ClientTransformIndexer.java
@@ -337,7 +337,7 @@ class ClientTransformIndexer extends TransformIndexer {
         logger.debug("[{}] updating persistent state of transform to [{}].", transformConfig.getId(), state.toString());
 
         // we might need to call the save state listeners, but do not want to stop rolling
-        doSaveState(state, ActionListener.wrap(r -> {
+        persistStateWithAutoStop(state, ActionListener.wrap(r -> {
             try {
                 if (saveStateListenersAtTheMomentOfCalling != null) {
                     ActionListener.onResponse(saveStateListenersAtTheMomentOfCalling, r);
@@ -362,13 +362,19 @@ class ClientTransformIndexer extends TransformIndexer {
         }));
     }
 
+    private void persistStateWithAutoStop(TransformState state, ActionListener<Void> listener) {
+        persistState(state, ActionListener.runBefore(listener, () -> {
+            if (state.getTaskState().equals(TransformTaskState.STOPPED)) {
+                context.shutdown();
+            }
+        }));
+    }
+
     /**
-     * Runs the persistent part of state storage
-     *
-     * Important: Might call into TransformTask, this should _only_ be called with an acquired indexer lock if and only if
-     * the lock for TransformTask has been acquired, too. See gh#75846
+     * Runs the persistence part of state storage
      */
-    private void doSaveState(TransformState state, ActionListener<Void> listener) {
+    @Override
+    protected void persistState(TransformState state, ActionListener<Void> listener) {
         // This could be `null` but the putOrUpdateTransformStoredDoc handles that case just fine
         SeqNoPrimaryTermAndIndex seqNoPrimaryTermAndIndex = getSeqNoPrimaryTermAndIndex();
 
@@ -380,10 +386,6 @@ class ClientTransformIndexer extends TransformIndexer {
             seqNoPrimaryTermAndIndex,
             ActionListener.wrap(r -> {
                 updateSeqNoPrimaryTermAndIndex(seqNoPrimaryTermAndIndex, r);
-                // for auto stop shutdown the task
-                if (state.getTaskState().equals(TransformTaskState.STOPPED)) {
-                    context.shutdown();
-                }
 
                 // Only do this clean up once, if it succeeded, no reason to do the query again.
                 if (oldStatsCleanedUp.compareAndSet(false, true)) {
@@ -427,9 +429,6 @@ class ClientTransformIndexer extends TransformIndexer {
                 } else {
                     logger.error(new ParameterizedMessage("[{}] updating stats of transform failed.", transformConfig.getId()), statsExc);
                     auditor.warning(getJobId(), "Failure updating stats of transform: " + statsExc.getMessage());
-                }
-                if (state.getTaskState().equals(TransformTaskState.STOPPED)) {
-                    context.shutdown();
                 }
                 listener.onFailure(statsExc);
             })

--- a/x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformIndexer.java
+++ b/x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformIndexer.java
@@ -39,6 +39,7 @@ import org.elasticsearch.xpack.core.transform.transforms.TransformConfig;
 import org.elasticsearch.xpack.core.transform.transforms.TransformIndexerPosition;
 import org.elasticsearch.xpack.core.transform.transforms.TransformIndexerStats;
 import org.elasticsearch.xpack.core.transform.transforms.TransformProgress;
+import org.elasticsearch.xpack.core.transform.transforms.TransformState;
 import org.elasticsearch.xpack.core.transform.transforms.TransformTaskState;
 import org.elasticsearch.xpack.core.transform.utils.ExceptionsHelper;
 import org.elasticsearch.xpack.transform.TransformServices;
@@ -163,6 +164,8 @@ public abstract class TransformIndexer extends AsyncTwoPhaseIndexer<TransformInd
     abstract void doDeleteByQuery(DeleteByQueryRequest deleteByQueryRequest, ActionListener<BulkByScrollResponse> responseListener);
 
     abstract void refreshDestinationIndex(ActionListener<RefreshResponse> responseListener);
+
+    abstract void persistState(TransformState state, ActionListener<Void> listener);
 
     public int getPageSize() {
         return pageSize;
@@ -691,14 +694,36 @@ public abstract class TransformIndexer extends AsyncTwoPhaseIndexer<TransformInd
 
         // in case the indexer isn't running, respond immediately
         if (state == IndexerState.STARTED && context.shouldStopAtCheckpoint() != shouldStopAtCheckpoint) {
-            context.setShouldStopAtCheckpoint(shouldStopAtCheckpoint);
+            IndexerState newIndexerState = IndexerState.STARTED;
+            TransformTaskState newtaskState = context.getTaskState();
+
+            // check if the transform is at a checkpoint, if so, we will stop it below
+            // otherwise we set shouldStopAtCheckpoint and wait
+            if (shouldStopAtCheckpoint && initialRun()) {
+                newIndexerState = IndexerState.STOPPED;
+                newtaskState = TransformTaskState.STOPPED;
+                logger.debug("[{}] transform is at a checkpoint, initiating stop.", transformConfig.getId());
+            } else {
+                context.setShouldStopAtCheckpoint(shouldStopAtCheckpoint);
+            }
+
+            final TransformState newTransformState = new TransformState(
+                newtaskState,
+                newIndexerState,
+                getPosition(),
+                context.getCheckpoint(),
+                context.getStateReason(),
+                getProgress(),
+                null,
+                shouldStopAtCheckpoint
+            );
 
             // because save state is async we need to block the call until state is persisted, so that the job can not
             // be triggered (ensured by synchronized)
             CountDownLatch latch = new CountDownLatch(1);
             logger.debug("[{}] persisting stop at checkpoint", getJobId());
 
-            doSaveState(IndexerState.STARTED, getPosition(), () -> { latch.countDown(); });
+            persistState(newTransformState, ActionListener.wrap(() -> latch.countDown()));
 
             if (latch.await(PERSIST_STOP_AT_CHECKPOINT_TIMEOUT_SEC, TimeUnit.SECONDS) == false) {
                 logger.error(
@@ -709,6 +734,12 @@ public abstract class TransformIndexer extends AsyncTwoPhaseIndexer<TransformInd
                     )
                 );
             }
+
+            // stop the transform if the decision was to stop it above
+            if (newtaskState.equals(TransformTaskState.STOPPED)) {
+                context.shutdown();
+            }
+
             return false;
         }
 

--- a/x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformIndexer.java
+++ b/x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformIndexer.java
@@ -19,6 +19,7 @@ import org.elasticsearch.action.search.SearchRequest;
 import org.elasticsearch.action.search.SearchResponse;
 import org.elasticsearch.action.support.IndicesOptions;
 import org.elasticsearch.common.breaker.CircuitBreakingException;
+import org.elasticsearch.common.logging.LoggerMessageFormat;
 import org.elasticsearch.common.util.CollectionUtils;
 import org.elasticsearch.core.Tuple;
 import org.elasticsearch.index.query.BoolQueryBuilder;
@@ -640,6 +641,119 @@ public abstract class TransformIndexer extends AsyncTwoPhaseIndexer<TransformInd
         auditor.info(transformConfig.getId(), "Received abort request, stopping transform.");
         logger.info("[{}] transform received abort request. Stopping indexer.", transformConfig.getId());
         context.shutdown();
+    }
+
+    @Override
+    protected void doSaveState(IndexerState indexerState, TransformIndexerPosition position, Runnable next) {
+        if (context.getTaskState() == TransformTaskState.FAILED) {
+            logger.debug("[{}] attempted to save state and stats while failed.", getJobId());
+            // If we are failed, we should call next to allow failure handling to occur if necessary.
+            next.run();
+            return;
+        }
+        if (indexerState.equals(IndexerState.ABORTING)) {
+            // If we're aborting, just invoke `next` (which is likely an onFailure handler)
+            next.run();
+            return;
+        }
+
+        // getting the listeners that registered till now, in theory a new listener could sneak in between this line
+        // and the next, however this is benign: we would store `shouldStopAtCheckpoint = true` twice which is ok
+        Collection<ActionListener<Void>> saveStateListenersAtTheMomentOfCalling = saveStateListeners.getAndSet(null);
+        boolean shouldStopAtCheckpoint = context.shouldStopAtCheckpoint();
+
+        // If we should stop at the next checkpoint, are STARTED, and with `initialRun()` we are in one of two states
+        // 1. We have just called `onFinish` completing our request, but `shouldStopAtCheckpoint` was set to `true` before our check
+        // there and now
+        // 2. We are on the very first run of a NEW checkpoint and got here either through a failure, or the very first save state call.
+        //
+        // In either case, we should stop so that we guarantee a consistent state and that there are no partially completed checkpoints
+        if (shouldStopAtCheckpoint && initialRun() && indexerState.equals(IndexerState.STARTED)) {
+            indexerState = IndexerState.STOPPED;
+            auditor.info(transformConfig.getId(), "Transform is no longer in the middle of a checkpoint, initiating stop.");
+            logger.info("[{}] transform is no longer in the middle of a checkpoint, initiating stop.", transformConfig.getId());
+        }
+
+        // This means that the indexer was triggered to discover changes, found none, and exited early.
+        // If the state is `STOPPED` this means that TransformTask#stop was called while we were checking for changes.
+        // Allow the stop call path to continue
+        if (hasSourceChanged == false && indexerState.equals(IndexerState.STOPPED) == false) {
+            if (saveStateListenersAtTheMomentOfCalling != null) {
+                ActionListener.onResponse(saveStateListenersAtTheMomentOfCalling, null);
+            }
+            next.run();
+            return;
+        }
+
+        TransformTaskState taskState = context.getTaskState();
+
+        if (indexerState.equals(IndexerState.STARTED) && context.getCheckpoint() == 1 && this.isContinuous() == false) {
+            // set both to stopped so they are persisted as such
+            indexerState = IndexerState.STOPPED;
+
+            auditor.info(transformConfig.getId(), "Transform finished indexing all data, initiating stop");
+            logger.info("[{}] transform finished indexing all data, initiating stop.", transformConfig.getId());
+        }
+
+        // If we are `STOPPED` on a `doSaveState` call, that indicates we transitioned to `STOPPED` from `STOPPING`
+        // OR we called `doSaveState` manually as the indexer was not actively running.
+        // Since we save the state to an index, we should make sure that our task state is in parity with the indexer state
+        if (indexerState.equals(IndexerState.STOPPED)) {
+            // If we are going to stop after the state is saved, we should NOT persist `shouldStopAtCheckpoint: true` as this may
+            // cause problems if the task starts up again.
+            // Additionally, we don't have to worry about inconsistency with the ClusterState (if it is persisted there) as the
+            // when we stop, we mark the task as complete and that state goes away.
+            shouldStopAtCheckpoint = false;
+
+            // We don't want adjust the stored taskState because as soon as it is `STOPPED` a user could call
+            // .start again.
+            taskState = TransformTaskState.STOPPED;
+        }
+
+        final TransformState state = new TransformState(
+            taskState,
+            indexerState,
+            position,
+            context.getCheckpoint(),
+            context.getStateReason(),
+            getProgress(),
+            null,
+            shouldStopAtCheckpoint
+        );
+        logger.debug("[{}] updating persistent state of transform to [{}].", transformConfig.getId(), state.toString());
+
+        // we might need to call the save state listeners, but do not want to stop rolling
+        persistStateWithAutoStop(state, ActionListener.wrap(r -> {
+            try {
+                if (saveStateListenersAtTheMomentOfCalling != null) {
+                    ActionListener.onResponse(saveStateListenersAtTheMomentOfCalling, r);
+                }
+            } catch (Exception onResponseException) {
+                String msg = LoggerMessageFormat.format("[{}] failed notifying saveState listeners, ignoring.", getJobId());
+                logger.warn(msg, onResponseException);
+            } finally {
+                next.run();
+            }
+        }, e -> {
+            try {
+                if (saveStateListenersAtTheMomentOfCalling != null) {
+                    ActionListener.onFailure(saveStateListenersAtTheMomentOfCalling, e);
+                }
+            } catch (Exception onFailureException) {
+                String msg = LoggerMessageFormat.format("[{}] failed notifying saveState listeners, ignoring.", getJobId());
+                logger.warn(msg, onFailureException);
+            } finally {
+                next.run();
+            }
+        }));
+    }
+
+    private void persistStateWithAutoStop(TransformState state, ActionListener<Void> listener) {
+        persistState(state, ActionListener.runBefore(listener, () -> {
+            if (state.getTaskState().equals(TransformTaskState.STOPPED)) {
+                context.shutdown();
+            }
+        }));
     }
 
     /**

--- a/x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformIndexer.java
+++ b/x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformIndexer.java
@@ -61,7 +61,7 @@ import java.util.stream.Stream;
 
 public abstract class TransformIndexer extends AsyncTwoPhaseIndexer<TransformIndexerPosition, TransformIndexerStats> {
 
-    private static final int PERSIST_STOP_AT_CHECKPOINT_TIMEOUT_SEC = 5;
+    private static final int PERSIST_STOP_AT_CHECKPOINT_TIMEOUT_SEC = 10;
 
     /**
      * RunState is an internal (non-persisted) state that controls the internal logic
@@ -696,7 +696,7 @@ public abstract class TransformIndexer extends AsyncTwoPhaseIndexer<TransformInd
             // because save state is async we need to block the call until state is persisted, so that the job can not
             // be triggered (ensured by synchronized)
             CountDownLatch latch = new CountDownLatch(1);
-            logger.debug("[{}] persisiting stop at checkpoint", getJobId());
+            logger.debug("[{}] persisting stop at checkpoint", getJobId());
 
             doSaveState(IndexerState.STARTED, getPosition(), () -> { latch.countDown(); });
 

--- a/x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformIndexer.java
+++ b/x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformIndexer.java
@@ -61,7 +61,7 @@ import java.util.stream.Stream;
 
 public abstract class TransformIndexer extends AsyncTwoPhaseIndexer<TransformIndexerPosition, TransformIndexerStats> {
 
-    private static final int PERSIST_STOP_AT_CHECKPOINT_TIMEOUT_SEC = 10;
+    private static final int PERSIST_STOP_AT_CHECKPOINT_TIMEOUT_SEC = 5;
 
     /**
      * RunState is an internal (non-persisted) state that controls the internal logic
@@ -755,7 +755,18 @@ public abstract class TransformIndexer extends AsyncTwoPhaseIndexer<TransformInd
         }
     }
 
-    synchronized void handleFailure(Exception e) {
+    /**
+     * Checks the given exception and handles the error based on it.
+     *
+     * In case the error is permanent or the number for failures exceed the number of retries, sets the indexer
+     * to `FAILED`.
+     *
+     * Important: Might call into TransformTask, this should _only_ be called with an acquired indexer lock if and only if
+     * the lock for TransformTask has been acquired, too. See gh#75846
+     *
+     * (Note: originally this method was synchronized, which is not necessary)
+     */
+    void handleFailure(Exception e) {
         logger.warn(new ParameterizedMessage("[{}] transform encountered an exception: ", getJobId()), e);
         Throwable unwrappedException = ExceptionsHelper.findSearchExceptionRootCause(e);
 

--- a/x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformIndexer.java
+++ b/x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformIndexer.java
@@ -682,7 +682,6 @@ public abstract class TransformIndexer extends AsyncTwoPhaseIndexer<TransformInd
         boolean shouldStopAtCheckpoint,
         ActionListener<Void> shouldStopAtCheckpointListener
     ) throws InterruptedException {
-
         // in case the indexer is already shutting down
         if (indexerThreadShuttingDown) {
             context.setShouldStopAtCheckpoint(shouldStopAtCheckpoint);
@@ -697,8 +696,9 @@ public abstract class TransformIndexer extends AsyncTwoPhaseIndexer<TransformInd
             IndexerState newIndexerState = IndexerState.STARTED;
             TransformTaskState newtaskState = context.getTaskState();
 
-            // check if the transform is at a checkpoint, if so, we will stop it below
-            // otherwise we set shouldStopAtCheckpoint and wait
+            // check if the transform is at a checkpoint, if so, we will shortcut and stop it below
+            // otherwise we set shouldStopAtCheckpoint, for this case the transform needs to get
+            // triggered, complete the checkpoint and stop
             if (shouldStopAtCheckpoint && initialRun()) {
                 newIndexerState = IndexerState.STOPPED;
                 newtaskState = TransformTaskState.STOPPED;
@@ -715,7 +715,7 @@ public abstract class TransformIndexer extends AsyncTwoPhaseIndexer<TransformInd
                 context.getStateReason(),
                 getProgress(),
                 null,
-                shouldStopAtCheckpoint
+                newIndexerState == IndexerState.STARTED
             );
 
             // because save state is async we need to block the call until state is persisted, so that the job can not

--- a/x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformTask.java
+++ b/x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformTask.java
@@ -311,6 +311,11 @@ public class TransformTask extends AllocatedPersistentTask implements SchedulerE
             return;
         }
 
+        if (context.shouldStopAtCheckpoint() == shouldStopAtCheckpoint) {
+            shouldStopAtCheckpointListener.onResponse(null);
+            return;
+        }
+
         // move the call to the generic thread pool, so we do not block the network thread
         getThreadPool().executor(ThreadPool.Names.GENERIC)
             .execute(() -> { getIndexer().setStopAtCheckpoint(shouldStopAtCheckpoint, shouldStopAtCheckpointListener); });

--- a/x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformTask.java
+++ b/x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformTask.java
@@ -300,6 +300,8 @@ public class TransformTask extends AllocatedPersistentTask implements SchedulerE
         boolean shouldStopAtCheckpoint,
         ActionListener<Void> shouldStopAtCheckpointListener
     ) {
+        // this should be called from the generic threadpool
+        assert Thread.currentThread().getName().contains(ThreadPool.Names.GENERIC);
         logger.debug(
             "[{}] attempted to set task to stop at checkpoint [{}] with state [{}]",
             getTransformId(),
@@ -316,9 +318,7 @@ public class TransformTask extends AllocatedPersistentTask implements SchedulerE
             return;
         }
 
-        // move the call to the generic thread pool, so we do not block the network thread
-        getThreadPool().executor(ThreadPool.Names.GENERIC)
-            .execute(() -> { getIndexer().setStopAtCheckpoint(shouldStopAtCheckpoint, shouldStopAtCheckpointListener); });
+        getIndexer().setStopAtCheckpoint(shouldStopAtCheckpoint, shouldStopAtCheckpointListener);
     }
 
     public synchronized void stop(boolean force, boolean shouldStopAtCheckpoint) {

--- a/x-pack/plugin/transform/src/test/java/org/elasticsearch/xpack/transform/transforms/TransformIndexerFailureHandlingTests.java
+++ b/x-pack/plugin/transform/src/test/java/org/elasticsearch/xpack/transform/transforms/TransformIndexerFailureHandlingTests.java
@@ -198,7 +198,7 @@ public class TransformIndexerFailureHandlingTests extends ESTestCase {
         @Override
         protected void doSaveState(IndexerState state, TransformIndexerPosition position, Runnable next) {
             assert state == IndexerState.STARTED || state == IndexerState.INDEXING || state == IndexerState.STOPPED;
-            next.run();
+            super.doSaveState(state, position, next);
         }
 
         @Override

--- a/x-pack/plugin/transform/src/test/java/org/elasticsearch/xpack/transform/transforms/TransformIndexerFailureHandlingTests.java
+++ b/x-pack/plugin/transform/src/test/java/org/elasticsearch/xpack/transform/transforms/TransformIndexerFailureHandlingTests.java
@@ -45,6 +45,7 @@ import org.elasticsearch.xpack.core.transform.transforms.TransformCheckpoint;
 import org.elasticsearch.xpack.core.transform.transforms.TransformConfig;
 import org.elasticsearch.xpack.core.transform.transforms.TransformIndexerPosition;
 import org.elasticsearch.xpack.core.transform.transforms.TransformIndexerStats;
+import org.elasticsearch.xpack.core.transform.transforms.TransformState;
 import org.elasticsearch.xpack.core.transform.transforms.TransformTaskState;
 import org.elasticsearch.xpack.transform.Transform;
 import org.elasticsearch.xpack.transform.TransformServices;
@@ -270,6 +271,11 @@ public class TransformIndexerFailureHandlingTests extends ESTestCase {
         @Override
         void doGetFieldMappings(ActionListener<Map<String, String>> fieldMappingsListener) {
             fieldMappingsListener.onResponse(Collections.emptyMap());
+        }
+
+        @Override
+        void persistState(TransformState state, ActionListener<Void> listener) {
+            listener.onResponse(null);
         }
     }
 

--- a/x-pack/plugin/transform/src/test/java/org/elasticsearch/xpack/transform/transforms/TransformIndexerStateTests.java
+++ b/x-pack/plugin/transform/src/test/java/org/elasticsearch/xpack/transform/transforms/TransformIndexerStateTests.java
@@ -256,6 +256,11 @@ public class TransformIndexerStateTests extends ESTestCase {
         void doGetFieldMappings(ActionListener<Map<String, String>> fieldMappingsListener) {
             fieldMappingsListener.onResponse(Collections.emptyMap());
         }
+
+        @Override
+        void persistState(TransformState state, ActionListener<Void> listener) {
+            listener.onResponse(null);
+        }
     }
 
     @Before

--- a/x-pack/plugin/transform/src/test/java/org/elasticsearch/xpack/transform/transforms/TransformIndexerStateTests.java
+++ b/x-pack/plugin/transform/src/test/java/org/elasticsearch/xpack/transform/transforms/TransformIndexerStateTests.java
@@ -259,6 +259,7 @@ public class TransformIndexerStateTests extends ESTestCase {
 
         @Override
         void persistState(TransformState state, ActionListener<Void> listener) {
+            persistedState = state;
             listener.onResponse(null);
         }
     }
@@ -307,6 +308,7 @@ public class TransformIndexerStateTests extends ESTestCase {
                 null,
                 threadPool,
                 auditor,
+                new TransformIndexerPosition(Collections.singletonMap("afterkey", "value"), Collections.emptyMap()),
                 new TransformIndexerStats(),
                 context
             );
@@ -322,6 +324,27 @@ public class TransformIndexerStateTests extends ESTestCase {
             }
         }
 
+        // test the case that the indexer is at a checkpoint already
+        {
+            AtomicReference<IndexerState> stateRef = new AtomicReference<>(IndexerState.STARTED);
+            TransformContext context = new TransformContext(TransformTaskState.STARTED, "", 0, mock(TransformContext.Listener.class));
+            final MockedTransformIndexer indexer = createMockIndexer(
+                config,
+                stateRef,
+                null,
+                threadPool,
+                auditor,
+                null,
+                new TransformIndexerStats(),
+                context
+            );
+            assertResponse(listener -> setStopAtCheckpoint(indexer, true, listener));
+            assertEquals(0, indexer.getSaveStateListenerCallCount());
+            // shouldStopAtCheckpoint should not be set, the indexer was started, however at a checkpoint
+            assertFalse(context.shouldStopAtCheckpoint());
+            assertFalse(indexer.getPersistedState().shouldStopAtNextCheckpoint());
+        }
+
         // lets test a running indexer
         AtomicReference<IndexerState> state = new AtomicReference<>(IndexerState.STARTED);
         {
@@ -332,6 +355,7 @@ public class TransformIndexerStateTests extends ESTestCase {
                 null,
                 threadPool,
                 auditor,
+                null,
                 new TransformIndexerStats(),
                 context
             );
@@ -361,6 +385,7 @@ public class TransformIndexerStateTests extends ESTestCase {
                 null,
                 threadPool,
                 auditor,
+                null,
                 new TransformIndexerStats(),
                 context
             );
@@ -391,6 +416,7 @@ public class TransformIndexerStateTests extends ESTestCase {
                 null,
                 threadPool,
                 auditor,
+                null,
                 new TransformIndexerStats(),
                 context
             );
@@ -434,6 +460,7 @@ public class TransformIndexerStateTests extends ESTestCase {
                 null,
                 threadPool,
                 auditor,
+                null,
                 new TransformIndexerStats(),
                 context
             );
@@ -509,6 +536,7 @@ public class TransformIndexerStateTests extends ESTestCase {
             null,
             threadPool,
             auditor,
+            null,
             new TransformIndexerStats(),
             context
         );
@@ -602,6 +630,7 @@ public class TransformIndexerStateTests extends ESTestCase {
         Consumer<String> failureConsumer,
         ThreadPool threadPool,
         TransformAuditor auditor,
+        TransformIndexerPosition initialPosition,
         TransformIndexerStats jobStats,
         TransformContext context
     ) {
@@ -620,7 +649,7 @@ public class TransformIndexerStateTests extends ESTestCase {
             checkpointProvider,
             config,
             state,
-            null,
+            initialPosition,
             jobStats,
             context
         );

--- a/x-pack/plugin/transform/src/test/java/org/elasticsearch/xpack/transform/transforms/TransformIndexerStateTests.java
+++ b/x-pack/plugin/transform/src/test/java/org/elasticsearch/xpack/transform/transforms/TransformIndexerStateTests.java
@@ -205,28 +205,11 @@ public class TransformIndexerStateTests extends ESTestCase {
 
         @Override
         protected void doSaveState(IndexerState state, TransformIndexerPosition position, Runnable next) {
-            persistedState = new TransformState(
-                context.getTaskState(),
-                state,
-                position,
-                context.getCheckpoint(),
-                context.getStateReason(),
-                getProgress(),
-                null,
-                context.shouldStopAtCheckpoint()
-            );
-
-            Collection<ActionListener<Void>> saveStateListenersAtTheMomentOfCalling = saveStateListeners.getAndSet(null);
-            try {
-                if (saveStateListenersAtTheMomentOfCalling != null) {
-                    saveStateListenerCallCount += saveStateListenersAtTheMomentOfCalling.size();
-                    ActionListener.onResponse(saveStateListenersAtTheMomentOfCalling, null);
-                }
-            } catch (Exception onResponseException) {
-                fail("failed to call save state listeners");
-            } finally {
-                next.run();
-            }
+            Collection<ActionListener<Void>> saveStateListenersAtTheMomentOfCalling = saveStateListeners.get();
+            saveStateListenerCallCount += (saveStateListenersAtTheMomentOfCalling != null)
+                ? saveStateListenersAtTheMomentOfCalling.size()
+                : 0;
+            super.doSaveState(state, position, next);
         }
 
         @Override

--- a/x-pack/plugin/transform/src/test/java/org/elasticsearch/xpack/transform/transforms/TransformIndexerTests.java
+++ b/x-pack/plugin/transform/src/test/java/org/elasticsearch/xpack/transform/transforms/TransformIndexerTests.java
@@ -239,7 +239,7 @@ public class TransformIndexerTests extends ESTestCase {
             assert state == IndexerState.STARTED || state == IndexerState.INDEXING || state == IndexerState.STOPPED;
 
             assertTrue(saveStateInProgress.compareAndSet(true, false));
-            next.run();
+            super.doSaveState(state, position, next);
         }
 
         @Override

--- a/x-pack/plugin/transform/src/test/java/org/elasticsearch/xpack/transform/transforms/TransformIndexerTests.java
+++ b/x-pack/plugin/transform/src/test/java/org/elasticsearch/xpack/transform/transforms/TransformIndexerTests.java
@@ -41,6 +41,7 @@ import org.elasticsearch.xpack.core.transform.transforms.TransformCheckpoint;
 import org.elasticsearch.xpack.core.transform.transforms.TransformConfig;
 import org.elasticsearch.xpack.core.transform.transforms.TransformIndexerPosition;
 import org.elasticsearch.xpack.core.transform.transforms.TransformIndexerStats;
+import org.elasticsearch.xpack.core.transform.transforms.TransformState;
 import org.elasticsearch.xpack.core.transform.transforms.TransformTaskState;
 import org.elasticsearch.xpack.transform.TransformServices;
 import org.elasticsearch.xpack.transform.checkpoint.CheckpointProvider;
@@ -265,6 +266,11 @@ public class TransformIndexerTests extends ESTestCase {
 
         public int getDeleteByQueryCallCount() {
             return deleteByQueryCallCount;
+        }
+
+        @Override
+        void persistState(TransformState state, ActionListener<Void> listener) {
+            listener.onResponse(null);
         }
     }
 

